pep-0548.rst

```diff
@@ -0,0 +1,338 @@
+PEP: 548
+Title: The "python" Command on Unix-Like Systems (2017 update)
+Version: $Revision$
+Last-Modified: $Date$
+Author: Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 31-Jul-2017
+Replaces: 394
+
+Abstract
+========
+
+This PEP provides an updated convention to help ensure that Python scripts can
+continue to be portable across ``*nix`` systems, regardless of the default
+version of the Python interpreter (i.e. the version invoked by the ``python``
+command), and without requiring platforms to provide parallel Python 2 & 3
+stacks indefinitely.
+
+* ``python2`` will refer to some version of Python 2.x.
+* ``python3`` will refer to some version of Python 3.x.
+* ``python`` should either be absent, refer to Python 2.7 *or* refer to
+  Python 3.6+.
+* if a ``python`` link is provided, the target Python interpreter should provide
+  the ``six`` compatibility library by default.
+* currently Python 2 only scripts should either be updated to be source
+  compatible with Python 3 or else to use ``python2`` in the shebang line.
+* Cross-platform applications should be updated to be independent of the default
+  Python version in the target deployment environment.
+
+Developed in 2017, these recommendations supersede the previous recommendations
+published in 2012 (PEP 394). They differ mainly in granting platform providers
+more scope to effectively manage the eventual removal of their Python 2.7
+stacks.
+
+
+Recommendations for Development of Standalone Scripts
+=====================================================
+
+* standalone scripts that are incompatible with Python 2.x should specifically
+  target ``python3``.
+* standalone scripts that are incompatible with Python 3.x, and do not need to
+  gain compatibility with platforms that only provide Python 3.x, should
+  specifically target ``python2`` (as this is available by default almost
+  everywhere that Python 2 is available, and is easily added as a symlink if
+  it is missing).
+* standalone scripts that want to support both older platforms where ``python``
+  refers to ``python2`` *and* newer platforms or virtual environments where it
+  refers to ``python3`` should be updated to run in the common subset of Python
+  2.7 & 3.6+ using either the `regular porting guide`_ in the Python reference
+  documentation or else the `conservative porting guide`_ developed as part of
+  Fedora's migration to using Python 3 as the system Python runtime.
```

Comment:
After writing https://fedoraproject.org/wiki/User:Ncoghlan/Default_python_module I'm actually wondering if I should put something in the migration notes about providing a custom error message that explains how to set `/usr/bin/python` to something sensible when it's absent by default, because the standard error message from Fedora's shebang handler is pretty unhelpful, and I assume we're far from alone in that :)